app/services/S3.scala

```diff
@@ -44,116 +60,111 @@ object S3 extends S3Client with StrictLogging {
-        Left(s"Error reading from S3: ${e.getMessage}")
+    }.mapError { e =>
+      logger.error(s"Error reading $objectSettings from S3: ${e.getMessage}", e)
+      S3GetObjectError(e)
     }
   }
 
-  private def put(objectSettings: S3ObjectSettings, data: String)(implicit ec: ExecutionContext): Either[String,Unit] = {
-    val bytes = data.getBytes(StandardCharsets.UTF_8)
-    val stream = new ByteArrayInputStream(bytes)
-
-    val metadata = new ObjectMetadata()
-    metadata.setContentLength(bytes.length)
-    // https://docs.fastly.com/en/guides/how-caching-and-cdns-work#surrogate-headers
-    objectSettings.cacheControl.foreach(metadata.setCacheControl)
-    objectSettings.surrogateControl.foreach(cc => metadata.addUserMetadata("surrogate-control", cc))
-
-    val request = new PutObjectRequest(
-      objectSettings.bucket,
-      objectSettings.key,
-      stream,
-      metadata
-    )
-
-    try {
-      s3Client.putObject(
-        if (objectSettings.publicRead) request.withCannedAcl(CannedAccessControlList.PublicRead)
-        else request
-      )
-    } catch { case NonFatal(e) =>
-      logger.error(s"Error writing $objectSettings to S3: ${e.getMessage}", e)
-    } finally {
-      stream.close()
-    }
-    Right(())
+  def update(data: RawVersionedS3Data): S3Action[Unit] = { objectSettings =>
+    effectBlocking(s3Client.getObjectMetadata(objectSettings.bucket, objectSettings.key).getVersionId)
+      .mapError(e => {
+        logger.error(s"Error getting object metadata for $objectSettings: ${e.getMessage}", e)
+        S3GetObjectError(e)
+      })
+      .flatMap(currentVersion => {
+        if (currentVersion == data.version) {
+          createOrUpdate(data.value)(objectSettings)
+        } else {
+          logger.warn(s"Cannot update S3 object $objectSettings because provided version (${data.version}) does not match latest version ($currentVersion)")
+          IO.fail(S3VersionMatchError)
+        }
+      })
   }
 
-  def update(objectSettings: S3ObjectSettings, data: RawVersionedS3Data)(implicit ec: ExecutionContext): Future[Either[String,RawVersionedS3Data]] = Future {
-    try {
-      val currentVersion = s3Client.getObjectMetadata(objectSettings.bucket, objectSettings.key).getVersionId
+  def createOrUpdate(data: String): S3Action[Unit] = { objectSettings =>
+    val bytes = data.getBytes(StandardCharsets.UTF_8)
 
-      if (currentVersion == data.version) {
-        put(objectSettings, data.value).map(_ => data)
-      } else {
-        logger.warn(s"Cannot update S3 object $objectSettings because provided version (${data.version}) does not match latest version ($currentVersion)")
-        Left(s"Can't save your settings because someone else has updated them since they were last fetched")
+    Task(new ByteArrayInputStream(bytes))
+      .bracket(stream => UIO(stream.close()))
+      .apply { stream =>
+
+        UIO.effectTotal {
```

Comment:
this block will not fail, but has mutable state while it's setting up the request object, and so it's good to handle it as a synchronous effect